### PR TITLE
Add cache

### PR DESCRIPTION
**Pull request by abergeron**
This adds a cache to kernel compiles which will hopefully speed-up certain use cases.  This cache is per-process (so multiple contexts can reuse the same compiled kernel if their GPUs are compatible).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,11 @@ execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_SOURCE_DIR}/l
 set(LIBRARY_OUTPUT_PATH ${CMAKE_SOURCE_DIR}/lib)
 
 set(CMAKE_OSX_ARCHITECTURES x86_64)
+set(CMAKE_C_VISIBILITY_PRESET hidden)
+set(CMAKE_VISIBILITY_INLINES_HIDDEN 1)
+if(POLICY CMP0063)
```
Comment (nouiz) [inline CMakeLists.txt]: Add a comment to tell that this is.

Comment (abergeron) [inline CMakeLists.txt]: This is an unfortunate consequence of using cmake.  The only reason this is here is to make it not spam warnings when building with newer versions.

I could add a generic comment explaining that, but I won't explain what each policy does individually, that would be too much work and is already easily searchable on the internet for the curious.

Comment (nouiz) [inline CMakeLists.txt]: that is good.

**Comment by abergeron** [comment]
Ok, done.

**Comment by abergeron** [comment]
What is happening with this?

**Comment by nouiz** [comment]
Just no time to review.

**Comment by abergeron** [comment]
I would put a bump on this since it's killing performance to not have it for the multi-gpu (I'm talking about 90s vs 8s for each epoch) and that is with nvrtc.

I just encountered that while testing my other PR about caching allocations.

**Comment by nouiz** [comment]
What is this 90s vs 8s?

**Comment by abergeron** [comment]
Execution time for one epoch (in seconds)

**Comment by abergeron** [comment]
The last commit here fixes a bunch of problems with compilation that somehow were still on the uploaded branch.  I blame a bad rebase (I think).

Apart from that it passes all tests.

**Comment by abergeron** [comment]
Last commit should fix the undefined symbol 'assert' on linux.

File: src/gpuarray_buffer_cuda.c [hunk 2]
```diff
@@ -910,6 +973,36 @@ static gpukernel *cuda_newkernel(void *c, unsigned int count,
           cuda_exit(ctx);
           return NULL;
         }
+        if (compile_cache == NULL)
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I would move this to before line 933, otherwise, sometimes the cache won't be defined.

Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: It's ok.  The check above is only to see if something is in the cache.  So if there is no cache, nothing's in there.

**Pull request merged**